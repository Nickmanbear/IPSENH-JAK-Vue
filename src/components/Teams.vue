<template>
  <div id="teams">
    <h1>Teams</h1>

    <input v-model="newTeamName" type="text" placeholder="Team name" @keydown.enter="createTeam">
    <button @click="createTeam()">Make team</button>

    <Team v-for="team in teams"
          :key="team.id"
          v-bind:team="team"
          v-bind:users="users"
          v-bind:username="myUsername"
          v-on:refresh="getTeams"/>
  </div>
</template>

<script>
// @ is an alias to /src
import axios from '@/axiosInstance';
import Team from '@/components/Team.vue';

export default {
  name: 'Teams',
  components: {
    Team,
  },
  data() {
    return {
      myUsername: '',
      teams: [],
      users: [],
      newTeamName: '',
      editingName: false,
    };
  },
  mounted() {
    this.getTeams();
    this.getUsers();
    this.getMyUsername();
  },
  methods: {
    getTeams() {
      axios.get('/team')
        .then((response) => {
          this.teams = response.data;
        });
    },
    getUsers() {
      axios.get('/user')
        .then((response) => {
          this.users = Object.keys(response.data)
            .map((key) => (
              { id: key, username: response.data[key] }
            ));
        });
    },
    getMyUsername() {
      axios.get('/user/me')
        .then((response) => {
          this.myUsername = response.data.username;
        });
    },
    createTeam() {
      const team = {
        id: 0,
        name: this.newTeamName,
      };
      this.newTeamName = '';

      axios.post('/team', team)
        .then((response) => {
          this.teams.push(response.data);
          this.$emit('refresh');
        });
    },
  },
};
</script>

<style lang="scss">
  #teams {
    position: fixed;
    top: 85px;
    right: 15px;
    z-index: 50;
    min-width: 320px;
    max-height: 80vh;
    overflow: scroll;
    border-radius: 4px;
    padding: 15px 20px;
    background-color: white;
    border: 1px solid white;
    -ms-overflow-style: none;
    box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);

    &::-webkit-scrollbar {
      display: none;
    }

    h1 {
      margin: 0;
      padding: 0;
    }

    button {
      border: none;
      font-size: 0.8em;
      border-radius: 4px;
      padding: 3px 5px;
      margin: 0;

      &:hover {
        cursor: pointer;
      }
    }

    input {
      border: none;
      background-color: #eee;
      font-family: Arial, serif;
      font-size: 1em;
      margin: 8px 5px 8px 0;
      border-radius: 4px;
      padding: 3px;
    }
  }
</style>
