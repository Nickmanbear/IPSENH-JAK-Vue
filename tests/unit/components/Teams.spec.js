import { shallowMount } from '@vue/test-utils';
import Teams from '@/components/Teams.vue';

jest.mock('axios', () => ({
  create: () => ({
    post: () => Promise.resolve({
      data: {
        id: 3,
        name: 'test team3',
        leader: {
          id: 1,
          username: 'teamleader',
        },
        members: [
          {
            id: 1,
            username: 'teamleader',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
    }),
    get: (url) => {
      let data;
      switch (url) {
        case '/team': data = [
          {
            id: 1,
            name: 'test team',
            leader: {
              id: 1,
              username: 'teamleader',
            },
            members: [
              {
                id: 1,
                username: 'teamleader',
              },
              {
                id: 2,
                username: 'member1',
              },
              {
                id: 3,
                username: 'member2',
              },
            ],
          },
          {
            id: 2,
            name: 'test team 2',
            leader: {
              id: 4,
              username: 'teamleader2',
            },
            members: [
              {
                id: 4,
                username: 'teamleader2',
              },
              {
                id: 2,
                username: 'member1',
              },
              {
                id: 3,
                username: 'member2',
              },
            ],
          },
        ];
          break;
        case '/user': data = {
          1: 'user1',
          2: 'user2',
          3: 'user3',
        };
          break;
        case '/user/me': data = {
          id: 1,
          username: 'teamleader',
        };
          break;
        default: data = [];
          break;
      }
      return Promise.resolve({
        data,
      });
    },
  }),
}));

describe('Teams', () => {
  let cmp;

  beforeEach(() => {
    cmp = shallowMount(Teams, {
      data() {
        return {
          myUsername: '',
          teams: [],
          users: [],
          newTeamName: '',
          editingName: false,
          selectedMemberName: '',
        };
      },
    });
  });

  it('should update newTeamName', () => {
    cmp.find('input').setValue('test teamName');
    expect(cmp.vm.newTeamName).toEqual('test teamName');
  });

  it('should get teams', async () => {
    await cmp.vm.getTeams();
    expect(cmp.vm.teams).toEqual([
      {
        id: 1,
        name: 'test team',
        leader: {
          id: 1,
          username: 'teamleader',
        },
        members: [
          {
            id: 1,
            username: 'teamleader',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
      {
        id: 2,
        name: 'test team 2',
        leader: {
          id: 4,
          username: 'teamleader2',
        },
        members: [
          {
            id: 4,
            username: 'teamleader2',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
    ]);
  });

  it('should get users', async () => {
    await cmp.vm.getUsers();
    expect(cmp.vm.users).toEqual([
      { id: '1', username: 'user1' },
      { id: '2', username: 'user2' },
      { id: '3', username: 'user3' },
    ]);
  });

  it('should get my username', async () => {
    await cmp.vm.getMyUsername();
    expect(cmp.vm.myUsername).toEqual('teamleader');
  });

  it('should add team', async () => {
    await cmp.vm.createTeam();
    expect(cmp.vm.teams).toEqual([
      {
        id: 1,
        name: 'test team',
        leader: {
          id: 1,
          username: 'teamleader',
        },
        members: [
          {
            id: 1,
            username: 'teamleader',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
      {
        id: 2,
        name: 'test team 2',
        leader: {
          id: 4,
          username: 'teamleader2',
        },
        members: [
          {
            id: 4,
            username: 'teamleader2',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
      {
        id: 3,
        name: 'test team3',
        leader: {
          id: 1,
          username: 'teamleader',
        },
        members: [
          {
            id: 1,
            username: 'teamleader',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
    ]);
  });

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot();
  });
});
