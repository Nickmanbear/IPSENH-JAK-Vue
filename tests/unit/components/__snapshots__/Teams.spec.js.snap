// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teams has the expected html structure 1`] = `
<div
  id="teams"
>
  <h1>
    Teams
  </h1>
   
  <input
    placeholder="Team name"
    type="text"
  />
   
  <button>
    Make team
  </button>
   
  <team-stub
    team="[object Object]"
    username="teamleader"
    users="[object Object],[object Object],[object Object]"
  />
  <team-stub
    team="[object Object]"
    username="teamleader"
    users="[object Object],[object Object],[object Object]"
  />
</div>
`;
